app/forkjoin/forkjoin.go

```diff
@@ -0,0 +1,214 @@
+// Note Join must only be called once, otherwise panics.
+type Join[I, O any] func() Results[I, O]
+
+// Work defines the work function signature workers will call.
+type Work[I, O any] func(ctx context.Context, input I) (output O, err error)
+
+// Results contains enqueued results.
+type Results[I, O any] <-chan Result[I, O]
+
+// Result contains the input and resulting output from the work function.
+type Result[I, O any] struct {
+	Input  I
+	Output O
+	Err    error
+}
+
+// Flatten blocks and returns all the outputs when all completed and
+// either the first non-context-cancelled error or context-cancelled
+// if all errors are context-cancelled (to return real failure reason).
+func (r Results[I, O]) Flatten() ([]O, error) {
+	var (
+		ctxErr   error
+		otherErr error
+		resp     []O
+	)
+	for result := range r {
+		resp = append(resp, result.Output)
+
+		if result.Err == nil {
+			continue
+		}
+
+		if errors.Is(result.Err, context.Canceled) && ctxErr == nil {
+			ctxErr = result.Err
+		}
+		if !errors.Is(result.Err, context.Canceled) && otherErr == nil {
+			otherErr = result.Err
+		}
+	}
+
+	if otherErr != nil {
+		return resp, otherErr
+	} else if ctxErr != nil {
+		return resp, ctxErr
+	}
+
+	return resp, nil
+}
+
+type options struct {
+	inputBuf int
+	workers  int
+	failFast bool
+}
+
+type Option func(*options)
+
+// WithWorkers returns an option configuring a forkjoin with w number of workers.
+func WithWorkers(w int) Option {
+	return func(o *options) {
+		o.workers = w
+	}
+}
+
+// WithInputBuffer returns an option configuring a forkjoin with an input buffer of length i.
+// Useful to prevent temporary blocking during calls to Fork.
+func WithInputBuffer(i int) Option {
+	return func(o *options) {
+		o.inputBuf = i
+	}
+}
+
+// WithFailFast returns an option configuring a forkjoin to stop execution on any error.
+// Active work function contexts are cancelled and no further inputs are
+// executed with remaining result errors being set to context cancelled.
+func WithFailFast() Option {
+	return func(o *options) {
+		o.failFast = true
+	}
+}
+
+// New returns a new forkjoin instance with generic input type I and output type O.
+// It provides an API for "doing work concurrently (fork) and then waiting for the results (join)".
+//
+// Usage:
+//   var workFunc := func(ctx context.Context, input MyInput) (MyResult, error) {
+//     ... do work
+//     return result, nil
+//   }
+//
+//   fork, join := forkjoin.New[MyInput,MyResult](ctx, workFunc)
+//   for _, in := range inputs {
+//     fork(in) // Note that calling fork AFTER join panics!
+//   }
+//
+//   resultChan := join()
+//   // Either read results from the channel as they appear
+//   for result := range resultChan { ... }
+//   // Or block until all results are complete and flatten
+//   results, firstErr := resultChan.Flatten()
+//
+func New[I, O any](ctx context.Context, work Work[I, O], opts ...Option) (Fork[I], Join[I, O]) {
```

Comment:
discussion: i'm wondering why we need the `fork-join` API? where is it intended to be used?

Reply:
It is similar to `errgroup.Group` or `sync.WaitGroup`, but it just supports `failfast` and result collection out of the box. It is a pattern we manually write in a lot in tests (in the code in a few places), so I thought we can create a library for that